common/reactive/src/main/java/io/helidon/common/reactive/BufferedEmittingPublisher.java

```diff
@@ -227,154 +392,99 @@ public int bufferSize() {
+    // - any changes to state are followed by maybeDrain, so the thread inside drain() can notice
+    //   that some state change has occurred:
+    //   - ignorePending
+    //     - error
+    //     - cancelled
+    //   - requested
+    //   - buffer contents
+    private void drain(Flow.Subscriber<? super T> sub) {
+        IllegalStateException ise = null;
+        for (int cont = 1; cont > 0; cont = contenders.addAndGet(-cont)) {
+            boolean terminateNow = ignorePending;
             try {
-                if (buffer.isEmpty() && emitter.emit(item)) {
-                    // Buffer drained, emit successful
-                    // saved time by skipping buffer
+                while (!terminateNow && requested.get() > emitted && !buffer.isEmpty()) {
+                    T item = buffer.poll();
+                    if (subscriber == null && sub != null) {
+                        sub.onNext(item);
+                    } else {
+                        subscriber.onNext(item);
+                    }
                     if (onEmitCallback != null) {
                         onEmitCallback.accept(item);
                     }
-                    return 0;
-                } else {
-                    // safe slower path thru buffer
-                    buffer.add(item);
-                    state.get().drain(this);
-                    return buffer.size();
-                }
-            } finally {
-                // If unbounded, check only once if buffer is empty
-                if (!safeToSkipBuffer && isUnbounded() && buffer.isEmpty()) {
-                    safeToSkipBuffer = true;
+                    emitted++;
+                    terminateNow = ignorePending;
                 }
+            } catch (RuntimeException re) {
+                abort(re);
             }
-        }
-    }
 
-    private int unboundedEmitOrBuffer(T item) {
-        // Not reachable unless unbounded req was made
-        // and buffer is empty
-        if (emitter.emit(item)) {
-            // Emit successful
-            if (onEmitCallback != null) {
-                onEmitCallback.accept(item);
+            if (terminateNow) {
+                cleanup();
             }
-            return 0;
-        } else {
-            // Emitter can be only in terminal state
-            // buffer for later retrieval by clearBuffer()
-            buffer.add(item);
-            return buffer.size();
-        }
-    }
 
-
-    private enum State {
-        READY_TO_EMIT {
-            @Override
-            <T> int emit(BufferedEmittingPublisher<T> publisher, T item) {
-                if (publisher.safeToSkipBuffer) {
-                    return publisher.unboundedEmitOrBuffer(item);
+            if (terminateNow || isCompleted()) {
+                try {
+                    // assert: cleanup in finally
+                    if (!cancelled) {
+                        cancelled = true;
```

Comment:
add `ignorePending` here, so non-compliant callers that happen to invoke `emit` after `complete` do not create more chaos - they will simply observe `terminateNow`, and behave like after `completeNow` - just `cleanup`, and never call anything on `subscriber` (which will be `null` by then)

Reply:
thx done